File: 515/FindLargestValueInEachTreeRow.scala
object FindLargestValueInEachTreeRow {
  class TreeNode(val _value: Int = 0) {
    var value: Int = _value
    var left: TreeNode = null
    var right: TreeNode = null
  }

  def largestValues(root: TreeNode): List[Int] = if (root == null) Nil else {
    val res = new collection.mutable.ListBuffer[Int]

    @scala.annotation.tailrec
    def loop(row: List[TreeNode]): Unit = if (row.nonEmpty) {
      val (nextRow, maxValue) = row.foldLeft((List.empty[TreeNode], Int.MinValue)) { case ((r, v), node) =>
          val t = v max node.value
          if (node.left != null && node.right != null)
            (node.left :: node.right :: r, t)
          else if (node.right != null)
            (node.right :: r, t)
          else if (node.left != null)
            (node.left :: r, t)
          else (r, t)
      }
      res += maxValue
      loop(nextRow)
    }

    loop(List(root))
    res.toList
  }

}

File: 515/find-largest-value-in-each-tree-row.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var largestValues = function (root) {
    let res = [];
    const getMax = (a, b = -Infinity) => Math.max(a, b);
    const dfs = (node, depth) => {
        if (!node) return;
        res[depth] = getMax(node.val, res[depth]);
        dfs(node.left, depth + 1);
        dfs(node.right, depth + 1);
    };

    dfs(root, 0);
    return res;
};